lib/class-wp-rest-block-types-controller.php

```diff
@@ -0,0 +1,376 @@
+
+		// Retrieve the list of registered collection query parameters.
+		$registered = $this->get_collection_params();
+		$namespace  = '';
+		if ( isset( $registered['namespace'] ) && ! empty( $request['namespace'] ) ) {
+			$namespace = $request['namespace'];
+		}
+
+		foreach ( $block_types as $slug => $obj ) {
+			$block_type = $this->prepare_item_for_response( $obj, $request );
+			if ( $namespace ) {
+				$pieces          = explode( '/', $obj->name );
+				$block_namespace = $pieces[0];
+				if ( $namespace !== $block_namespace ) {
+					continue;
+				}
+			}
+			$data[ $obj->name ] = $this->prepare_response_for_collection( $block_type );
+		}
+
+		return rest_ensure_response( $data );
+	}
+
+	/**
+	 * Checks if a given request has access to read a block type.
+	 *
+	 * @param WP_REST_Request $request Full details about the request.
+	 *
+	 * @return WP_Error|bool True if the request has read access for the item, WP_Error object otherwise.
+	 */
+	public function get_item_permissions_check( $request ) {
+		$check = $this->check_read_permission();
+		if ( ! $check ) {
+			return new WP_Error( 'rest_cannot_read_block_type', __( 'Cannot view block type.', 'gutenberg' ), array( 'status' => rest_authorization_required_code() ) );
+		}
+		$block_name = sprintf( '%s/%s', $request['namespace'], $request['name'] );
+		$block_type = $this->get_block( $block_name );
+		if ( is_wp_error( $block_type ) ) {
+			return $block_type;
+		}
+
+		return true;
+	}
+
+	/**
+	 * Checks whether a given block type should be visible.
+	 *
+	 * @return WP_Error|bool True if the block type is visible, otherwise false.
+	 */
+	protected function check_read_permission() {
+		if ( ! current_user_can( 'edit_posts' ) ) {
+			return new WP_Error( 'rest_cannot_view', __( 'Sorry, you are not allowed to manage block types.', 'gutenberg' ), array( 'status' => rest_authorization_required_code() ) );
+		}
+
+		return true;
+	}
+
+	/**
+	 * Get the block, if the name is valid.
+	 *
+	 * @param string $name Block name.
+	 * @return WP_Block_Type|WP_Error Block type object if name is valid, WP_Error otherwise.
+	 */
+	protected function get_block( $name ) {
+		$block_type = $this->registry->get_registered( $name );
+		if ( empty( $block_type ) ) {
+			return new WP_Error( 'rest_block_type_invalid', __( 'Invalid block type.', 'gutenberg' ), array( 'status' => 404 ) );
+		}
+
+		return $block_type;
+	}
+
+	/**
+	 * Retrieves a specific block type.
+	 *
+	 * @param WP_REST_Request $request Full details about the request.
+	 *
+	 * @return WP_Error|WP_REST_Response Response object on success, or WP_Error object on failure.
+	 */
+	public function get_item( $request ) {
+		$block_name = sprintf( '%s/%s', $request['namespace'], $request['name'] );
+		$block_type = $this->get_block( $block_name );
+		if ( is_wp_error( $block_type ) ) {
+			return $block_type;
+		}
+		$data = $this->prepare_item_for_response( $block_type, $request );
+
+		return rest_ensure_response( $data );
```

Comment:
This can be removed, `prepare_item_for_response` returns a `WP_REST_Response` object.

Reply:
Isn't it just better to always pass this through `rest_ensure_response`? What if someone returned a `WP_Error` from `rest_prepare_block_type` filter?

Reply:
We [decided not to support](https://core.trac.wordpress.org/ticket/40580) that, but even if they did, `rest_ensure_response` passes through a `WP_Error` instance.

Reply:
How is this different from than [this](https://github.com/WordPress/wordpress-develop/blob/master/src/wp-includes/rest-api/endpoints/class-wp-rest-post-types-controller.php#L161-L163) or [this](https://github.com/WordPress/wordpress-develop/blob/master/src/wp-includes/rest-api/endpoints/class-wp-rest-comments-controller.php#L427-L428) or [this](https://github.com/WordPress/wordpress-develop/blob/master/src/wp-includes/rest-api/endpoints/class-wp-rest-posts-controller.php#L505-L506)?

Reply:
Core is inconsistent. In other routes there are plenty of  instances of core not applying `rest_ensure_response` before operating on it. If you look at the `get_items` route for any handler, it doesn't call `rest_ensure_response` before preparing for collection.

If you want to keep it here, keep it. But I'd really like to remove them from core as well. They don't actually provide stability, but in fact decrease it, since if someone returns a non `WP_REST_Response|WP_Error` instance from one of those filters, core will silently accept it and the developer won't think they've done something wrong. But then any other function that is hooked to that filter will have to deal with a potentially invalid object. And of course right now if someone does return an invalid value from the `prepare_item_for_response` core will crash anyways.

Reply:
I think for now, we should leave this. @TimothyBJacobs Can you create a ticket regarding removing them in core? Once it has been decided there / this PR goes into core, we can remove then. This is not core yet. The endpoint will likely change a little when it comes a core patch..